### PR TITLE
multi: use the new htlc v2 scripts for loop in/out with fallback to htlc v1 for old clients

### PR DESCRIPTION
**Pull request by bhandras**
This PR builds (and is rebased) on top of https://github.com/lightninglabs/loop/pull/255 (first two commits)

The PR integrates the new htlc v2 script to loop in/out while keeping compatibility for old clients. Compatibility is ensured by keeping track of the protocol version for swaps, with old swaps defaulting the the latest released protocol.

### CODE REVIEW

**Comment by bhandras** [comment]
Looking for concept ACK at this point as the test coverage is not fully extended yet.

File: loopdb/codec.go
Comment (carlaKC) [inline loopdb/codec.go:18]: Server side we went for a separate enum for DB when storing protocol version. Reasoning was that a value not written in the DB isn't actually the same as a legacy rpc protcol version (they may be on preimage push version, for example, and we just weren't storing at that time). I was also concerned about a screw up on rpc level leaking through to the DB, although in practice it's unlikely to happen.

File: loopdb/store.go
Comment (carlaKC) [inline loopdb/store.go:85]: So if protocol version is not written for a swap we use the latest version, couldn't that report a legacy client which isn't on this version yet as supporting it?

File: loopdb/store.go
Comment (halseth) [inline loopdb/store.go:389]: Do we need to pass in the default version? Could be more readable to just set the default first, then only call the method if anything is found in the bucket.

File: loopdb/codec.go
Comment (bhandras) [reply loopdb/codec.go:18]: Please note that using the proto generated enum is preexisting in the loop code.

In this change we don't fall back to the legacy version, but instead to `ProtocolVersion_USER_EXPIRY_LOOP_OUT` (which is the previous one) as the exact previous version does not matter as we only serialize for new swaps, and in that case the saved version will be `ProtocolVersion_HTLC_V2` (see the commit changing the current version).

The reasoning for not creating an extra enum wrapper over the proto generated enums is that it adds extra complexity where imo is not needed.

So to summarize:
- for old swaps: protocol version will be only used to decide between htlc v1/v2 scripts
- for newly created swaps: protocol version will be `ProtocolVersion_HTLC_V2` (see commit changing the current version for completeness)

File: loopdb/store.go
Comment (bhandras) [reply loopdb/store.go:85]: For all we know this is undecidable, but also because we don't actually serialize or send the version for swaps that don't have it serialized, it's just a matter of preference. We can safely change it to legacy as well, as it really only is used to decide between the two script versions.

File: loopdb/store.go
Comment (bhandras) [reply loopdb/store.go:389]: The reason this is passed is to avoid repeating the same checking and defaulting code where we unmarshal the protocol version. Can slice up if you wish but imo it's cleaner to keep together. WDYT?

File: loopdb/store.go
Comment (carlaKC) [inline loopdb/store.go:599]: I know I juust did this with labels, 😅 , but now that we have 2 fields that are common the the swap types but not in the swap contract, it may be nice to have a common function which takes the bucket and adds these fields. If we choose not to migrate swap contract, and continue with adding new keys it will be nice to have them in the same place. n/b

File: loopin.go
Comment (carlaKC) [inline loopin.go:180]: nit, linebreak is before 80

File: loopdb/store.go
Comment (halseth) [reply loopdb/store.go:433]: I mean
```go
protocolVersion := default
versionBytes := swapBucket.Get(protocolVersionKey)
if versionBytes != nil {
   protocolVersion, err = UnmarshalProtocolVersion(versionBytes)
   if err != nil {
      return err
   }
}
```
the checks on the passed version bytes could still be performed.

File: loopdb/codec.go
Comment (bhandras) [reply loopdb/codec.go]: I strongly think it'd better to use proto generated enums but I also decided to instead go with the more common opinion and translate between RPC and DB enums to also keep the code more compatible with the server side. PTAL

File: loopdb/store.go
Comment (bhandras) [reply loopdb/store.go]: Decided to instead go with the enum translation and add the the unrecorded similarly as on the server side. PTAL

File: loopdb/store.go
Comment (bhandras) [reply loopdb/store.go]: done

File: loopin.go
Comment (bhandras) [reply loopin.go:180]: done

File: loopin_test.go
Comment (carlaKC) [inline loopin_test.go:355]: If we are setting the script version in the test, why don't we just pass in the htlc version? Since we're not actually testing our version -> htlc type mapping, just duplicating it in a test.

File: swap.go
Comment (carlaKC) [inline swap.go:58]: Won't this set HtlcV2 for s.protocolVersion = `ProtocolVersionUnrecorded`? Which would mean any swap that is resumed after upgrade would have the wrong htlc type.

File: client.go
Comment (bhandras) [reply client.go]: done

File: loopdb/store.go
Comment (bhandras) [reply loopdb/store.go]: done

File: swap.go
Comment (bhandras) [reply swap.go]: Good catch! This was left when changed to db protocol version from proto protocol version.

File: loopin_test.go
Comment (bhandras) [reply loopin_test.go]: done, ptal

File: swap.go
Comment (carlaKC) [inline swap.go]: still on their own newlines?

File: swap/htlc.go
Comment (carlaKC) [inline swap/htlc.go:36]: sorry to keep nitting thing. "improved or V2" sounds like this represents two different things to me. 
My previous comment was more to briefly describe how the script is different form V1, but non-blocking.

File: loopdb/store.go
Comment (bhandras) [reply loopdb/store.go:602]: I was thinking about this, and I'm unsure if that'd make the code more readable or harder to follow. Maybe I tend to be more on the "keeping them separate" side to keep them explicit rather then hide behind a function.

File: swap.go
Comment (halseth) [inline swap.go:78]: nit: might as well let this method return the script directly now? Maybe we don't need to define `swap.ScriptVersion`

File: loopdb/store.go
Comment (carlaKC) [reply loopdb/store.go:602]: indeed, was on the fence about it myself so let's leave as is.

**Comment by carlaKC** [comment]
New changes LGTM, having protocol version in `SwapContract` cleans things up nicely.

File: looprpc/version.go
Comment (joostjager) [inline looprpc/version.go]: Isn't it better to move it to the `loopdb` package, because that are the types that we use internally?

File: swap.go
Comment (joostjager) [reply swap.go:78]: That would mix protocol version knowledge into the htlc type. I think it is better like it is.

File: loopout.go
Comment (joostjager) [inline loopout.go:235]: This PR really has two changes: store protocol version and add htlc v2 (dependent on stored protocol version). You could order the commits in the same way.

File: loopin_test.go
Comment (joostjager) [inline loopin_test.go:285]: No loop out test?

File: swap/htlc.go
Comment (bhandras) [reply swap/htlc.go:34]: I'd consider constants to be always better than magic numbers (even if in this case these are not magic in any sense).

File: loopdb/protocol_version.go
Comment (bhandras) [reply loopdb/protocol_version.go:41]: Yeah, we don't need, but it's nicer to keep an unrecorded "version". The first incarnation of this PR simply used the protocol version enum generated from the proto definition and I fixed anything that's not recorded to the previous protocol version. That isn't entirely correct either as we have no information about which version of the client started that swap. One could say it's not a problem either as we didn't use the version before for swaps, but it's not nice either.
Now that the separation of proto generated and DB protocol version is done, I'd say we can keep as is as at least this is more extendable whereas the proto generated is fixed.

File: looprpc/version.go
Comment (bhandras) [reply looprpc/version.go]: The reason it's in the `looprpc` package is because the constant's source is from proto and also there's no circular dependency that is created this way. I'd say it's better to keep it here.

File: loopdb/protocol_version.go
Comment (joostjager) [reply loopdb/protocol_version.go:41]: Ah yes indeed, also the client can have older swaps in the db.

File: loopout.go
Comment (bhandras) [reply loopout.go:227]: makes sense, ptal

File: looprpc/version.go
Comment (bhandras) [reply looprpc/version.go]: Also decided to remove the `ConvertProtocolVersion` function as we only used it to convert the compile time constant. Added a test to keep constants in sync and sane though.

File: loopdb/protocol_version.go
Comment (joostjager) [inline loopdb/protocol_version.go:31]: Fix in previous pr?

File: swap.go
Comment (joostjager) [inline swap.go:69]: Can be an early return

File: loopdb/store.go
Comment (bhandras) [reply loopdb/store.go:435]: Do you mean something like this:
```
func readProtocolVersion(swapBucket) (ProtocolVersion, error) {
        return UnmarshalProtocolVersion(swapBucket.Get(protocolVersionKey)
}
``` 
Would this make the code any more readable or simpler?

File: swap_server_client.go
Comment (bhandras) [reply swap_server_client.go:331]: We don't use the protocol version for these RPCs, they're just forwarded, as we only care about the hash preimage. We can still update as it's more consistent. I'm ok with updating actually. @carlaKC ?

File: swap_server_client.go
Comment (bhandras) [reply swap_server_client.go:268]: See my answer below.

File: loopdb/protocol_version.go
Comment (bhandras) [reply loopdb/protocol_version.go:31]: done

File: swap.go
Comment (bhandras) [reply swap.go]: done

File: loopdb/protocol_version.go
Comment (joostjager) [inline loopdb/protocol_version.go]: Somehow I keep thinking of a hard cast here instead of always updating two fields...

File: loopdb/store.go
Comment (joostjager) [reply loopdb/store.go:435]: I was thinking
```
func readCommonSwapProperties(&contract, swapBucket) error {
   contract.ProtocolVersion, err = UnmarshalProtocolVersion(
	swapBucket.Get(protocolVersionKey),
   )
   return err
}
```

But with only one property it is not very clearly better. It does make it easier to add the second shared property.

Comment (joostjager) [reply loopdb/store.go:435]: It is non-blocking

File: swap.go
Comment (joostjager) [inline swap.go:77]: It looks like the quote call still always uses htlc v1. I don't think it matters much, but then we need to change it when we eventually phase out v1.

Comment (bhandras) [reply swap.go:77]: Indeed, fixed.

File: loopdb/protocol_version.go
Comment (bhandras) [reply loopdb/protocol_version.go]: agree, fixed

File: loopdb/store.go
Comment (bhandras) [reply loopdb/store.go:435]: I honestly think it's better to keep as is, as it's easier to follow. Unless you insist you like the extracted method better, I'd keep it.

**Review by joostjager — APPROVED** [review]
LGTM but blocked on server release

File: swap/htlc.go
Comment (bhandras) [reply swap/htlc.go]: fixed

File: loopdb/codec_test.go
Comment (bhandras) [reply loopdb/codec_test.go:14]: :rofl:

File: swap/htlc.go
Comment (bhandras) [reply swap/htlc.go:182]: I'll update this before merge. (sparing a few rebases)

**Comment by cfromknecht** [comment]
@bhandras latest pass looks very much there 🎉

there was one lingering comment i had from the preceding PR which maybe we want to tack on here? 
https://github.com/lightninglabs/loop/pull/255#discussion_r478624667

**Pull request merged**
